himbaechel/uarch/gowin/gowin.cc

```diff
@@ -0,0 +1,400 @@
+    }
+    if (bel_type == id_VCC) {
+        return cell_type == id_GOWIN_VCC;
+    }
+    return (bel_type == cell_type);
+}
+
+void GowinImpl::assign_cell_info()
+{
+    fast_cell_info.resize(ctx->cells.size());
+    for (auto &cell : ctx->cells) {
+        CellInfo *ci = cell.second.get();
+        auto &fc = fast_cell_info.at(ci->flat_index);
+        if (is_lut(ci)) {
+            fc.lut_f = ci->getPort(id_F);
+            continue;
+        }
+        if (is_dff(ci)) {
+            fc.ff_d = ci->getPort(id_D);
+            fc.ff_clk = ci->getPort(id_CLK);
+            fc.ff_ce = ci->getPort(id_CE);
+            for (IdString port : {id_SET, id_RESET, id_PRESET, id_CLEAR}) {
+                fc.ff_lsr = ci->getPort(port);
+                if (fc.ff_lsr != nullptr) {
+                    break;
+                }
+            }
+            continue;
+        }
+        if (is_alu(ci)) {
+            fc.alu_sum = ci->getPort(id_SUM);
+            continue;
+        }
+    }
+}
+
+// placement validation
+bool GowinImpl::slice_valid(int x, int y, int z) const
+{
+    const CellInfo *lut = ctx->getBoundBelCell(ctx->getBelByLocation(Loc(x, y, z * 2)));
+    const CellInfo *ff = ctx->getBoundBelCell(ctx->getBelByLocation(Loc(x, y, z * 2 + 1)));
+    const CellInfo *alu = ctx->getBoundBelCell(ctx->getBelByLocation(Loc(x, y, z + BelZ::ALU0_Z)));
+    const CellInfo *ramw =
+            (z == 4 || z == 5) ? ctx->getBoundBelCell(ctx->getBelByLocation(Loc(x, y, BelZ::RAMW_Z))) : nullptr;
+
+    if (alu && lut) {
+        return false;
+    }
+
+    if (ramw) {
+        if (alu || ff || lut) {
+            return false;
+        }
+        return true;
+    }
+
+    // check for ALU/LUT in the adjacent cell
+    int adj_lut_z = (1 - (z & 1) * 2 + z) * 2;
+    int adj_alu_z = adj_lut_z / 2 + BelZ::ALU0_Z;
+    const CellInfo *adj_lut = ctx->getBoundBelCell(ctx->getBelByLocation(Loc(x, y, adj_lut_z)));
+    const CellInfo *adj_ff = ctx->getBoundBelCell(ctx->getBelByLocation(Loc(x, y, adj_lut_z + 1)));
+    const CellInfo *adj_alu = ctx->getBoundBelCell(ctx->getBelByLocation(Loc(x, y, adj_alu_z)));
+
+    if ((alu && (adj_lut || (adj_ff && !adj_alu))) || ((lut || (ff && !alu)) && adj_alu)) {
+        return false;
+    }
+
+    // if there is DFF it must be connected to this LUT or ALU
+    if (ff) {
+        const auto &ff_data = fast_cell_info.at(ff->flat_index);
+        if (lut) {
+            const auto &lut_data = fast_cell_info.at(lut->flat_index);
+            if (ff_data.ff_d != lut_data.lut_f) {
+                return false;
+            }
+        }
+        if (alu) {
+            const auto &alu_data = fast_cell_info.at(alu->flat_index);
+            if (ff_data.ff_d != alu_data.alu_sum) {
+                return false;
+            }
+        }
+        if (adj_ff) {
+            // DFFs must be same type or compatible
+            if (ff->type != adj_ff->type && ((ff->type.in(id_DFFS) && !adj_ff->type.in(id_DFFR)) ||
```

Comment:
maybe it would be neater to put this in a `ffs_compatible` inline function or similar?

Reply:
> maybe it would be neater to put this in a `ffs_compatible` inline function or similar?
This is really a one-time piece ... well, I'll move it tomorrow.

Reply:
> maybe it would be neater to put this in a `ffs_compatible` inline function or similar?

Done